services/placement/service/placementservice.go

```diff
@@ -84,29 +77,47 @@ func (ps placementService) BuildInitialPlacement(
 		}
 	}
 
-	return p, ps.validateAndSaveSnapshot(p)
+	if err := placement.Validate(p); err != nil {
+		return nil, err
+	}
+
+	// TODO(chaowant) this will be removed once the m3db nodes start to mark shards as available
+	p, err = markAllShardsAsAvailable(p)
+	if err != nil {
+		return nil, err
+	}
+
+	return p, ps.ss.SetIfNotExist(ps.service, p)
 }
 
 func (ps placementService) AddReplica() (services.ServicePlacement, error) {
-	var p services.ServicePlacement
-	var err error
-	if p, err = ps.Placement(); err != nil {
+	p, v, err := ps.ss.Placement(ps.service)
+	if err != nil {
 		return nil, err
 	}
 
 	if p, err = ps.algo.AddReplica(p); err != nil {
 		return nil, err
 	}
 
-	return p, ps.validateAndSaveSnapshot(p)
+	if err := placement.Validate(p); err != nil {
+		return nil, err
+	}
+
+	// TODO(chaowant) this will be removed once the m3db nodes start to mark shards as available
```

Comment:
For this case, let's just assume for now that nodes will make them as available.  That is how we intend to launch with this feature anyhow.  i.e. perhaps remove the `markAllShardsAsAvailable(...)` call for this case.